Wrappers/Python/cil/optimisation/functions/StochasticGradientAlgorithmFunction.py

```diff
@@ -0,0 +1,120 @@
+from cil.optimisation.functions import Function, SumFunction
+import numpy as np
+
+class StochasticGradientAlgorithmFunction(SumFunction):
+    
+    '''Class for use as objective function in Stochastic Gradient type algorithms such as GD
+    
+    Parameters:
+    -----------
+
+
+
+    '''
+    
+    # name suggestions: CustomGradientFunction, 
+    #                   A name with the subsets word
+    #     
+
+    
+    def __init__(self, functions, **kwargs):
+        # should not have docstring
+        super(StochasticGradientAlgorithmFunction, self).__init__(*functions)
+        
+    @property
+    def num_subsets(self):
+        return len(self.functions)
+        
+    def _full_gradient(self, x, out=None):
+        '''Return full gradient'''
+        return super(StochasticGradientAlgorithmFunction, self).gradient(x, out=out)
+        
+    def gradient(self, x, out=None):
+        """        
+            maybe calls _approx_grad inside        
+        """
+        raise NotImplemented
+
+
+    def subset_gradient(self, x, subset_num,  out=None):
+
+        return self.functions[subset_num].gradient(x)
+
+
+    def next_subset(self):
+
+        raise NotImplemented
+
+
+class SAGA_Function(StochasticGradientAlgorithmFunction):
+
+    def __init__(self, functions):
+
+        self.gradients_allocated = False
```

Comment:
See review regarding `memory_init()` and `gradients_allocated` but furthermore, all `StochasticGradientAlgorithmFunction` derived classes will require this variable.

Reply:
I'm not sure. In some cases we don't want to allocate memory, for stochastic gradient descent for example.

Reply:
The variable could still exist even for SGD (that is, in the parent class), it's just that the value will never change to True.  The question though is if that would be confusing (I'm assuming it wouldn't, but it might)

Reply:
Agreed. I don't think it would be confusing.